sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+} // namespace sycl::ext::oneapi
+```
+
+## New `async_group_copy` function.
+Currently the `nd_item` and `group` classes have the member function
+`async_work_group_copy`. This will be deprecated in favor of a
+`async_group_copy` that will have the group type as a template parameter, much
+like `group_barrier`.
+
+`async_group_copy` methods would be valid for groups `group and `sub_group`.
+For subgroup, a new bitmasked implementation will also be required.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride, sycl::sub_group_mask mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Change to `device_event::wait` function
+`device_event::wait` will be modified to take the group that will be synchonised in the wait.
+This will be valid for the same groups as `async_group_copy` i.e. `group` or `sub_group`.
+An overload with a `sub_group_mask` parameter will be required for `sub_group`.
+
+```c++
+namespace sycl {
+  class device_event {
+  ...
+  public:
+    template <typename Group>
+    void ext_oneapi_wait(Group group) noexcept;
+    void ext_oneapi_wait(sycl::sub_group group, sycl::sub_group_mask mask) noexcept;
+  ...
+  };
+}  // namespace sycl
+```
+
+## Open Questions
+
+* Do the `sub_group_mask` overloads of functions also need the `sub_group` class as an argument, since this is implicit.
```

Comment:
I suspect this is important to some library-only implementations, because a `sub_group` object may store important state that the work-items use to synchronize.  If that state had to be stored in `sub_group_mask` as well, it couldn't be implemented as a simple type storing just bits.  @illuhad?

Reply:
I don't think it much of a problem anyway, so we should probably just leave it in.